### ISSUE TITLE
Moving issues to their repository?

### ISSUE THREAD

**Issue by Martin-P** [zendframework/zendframework#7638]
Hi all,

Zend Framework is split up into several separate repositories for a while now. At this moment this repository contains 475 open issues. Are there any plans to migrate the issues to their matching repository as well so they won't be forgotten and overlooked?

**Comment by Maks3w** [comment]
Old issues are reviewed from time to time. New issues (after 2.5 split) may closed pointing to the correct issue tracker.

**Comment by Martin-P** [comment]
@Maks3w thanks for your fast response. So the issues are reviewed, but my question was if they are being migrated to their repositories? I got 2 open issues (#5826 and #7172) and looking at the response in those issues, the last review was over 8 months ago or didn't I understand your answer correctly? Migrating them to their repositories would probably make them much more visible.

Perhaps @weierophinney can tell us more about it?

**Comment by Maks3w** [comment]
1. This project is made by volunteers 

2. Issues are triaged so broken code is on top and questions/discussions are on the bottom

3. Not everybody knows everything. I'm only focused in an small set of components and QA

4. Move issues from one place to another does not fix nothing. I told because lots of issues was migrated from Jira and wasn't attended at all

**Comment by Martin-P** [comment]
> This project is made by volunteers

I am fully aware of that. FWIW: I don't get payed either for contributing to ZF2 :wink: 

> Not everybody knows everything. I'm only focused in an small set of components and QA

No problem, I don't expect everybody knowing everything. You say your focus is limited to a small set of components and QA. Perhaps you can elaborate why you closed this issue?

> Move issues from one place to another does not fix nothing. I told because lots of issues was migrated from Jira and wasn't attended at all

The reason for that was quite simple, because I faced it myself. The issues were moved to GitHub and I had to join GitHub to access my own issues and subscribe to my issues to receive any status notifications. Furthermore the issues were pretty garbled because of the migration: alignment fell short and code blocks were barely readible.

Fortunately those problems will not be there when moving issues inside GitHub, because it is the same platform with the same users and references.

**Comment by adamlundrigan** [comment]
Matthew and I spoke briefly about this at ZendCon and there are two possibiltiies:

1. Auto-close all open issues with a message asking the ticket author to re-open the ticket under the correct repository if the issue is still relevant. 

2. Manually review each open issue, closing the ones that are incomplete (not reproducible, insufficient detail, etc) and migrating the remaining ones to the correct repository.  

My vote is on (2), but it's a major endeavour to review the 475 issues that are open, determine if each is valid and then copy-and-paste to the appropriate repository.  If all the issues are tagged with the component then we could use a script to do the migration automatically, but that still leaves a lot of manual review work.

I've been slowly doing a first-pass through the open issues closing any that have insufficient detail or provide a reproducing case that now works against master.

**Comment by weierophinney** [comment]
Here's what I'm thinking:

- use a script to close issues older than X months, with a final comment indicating reporters should open a new issue/PR on the relevant repo if this is still a problem. We can do this *now*.
- Call to action to the CR team to go through the remaining open issues and ensure they are labeled with component-appropriate tags.
- Script to migrate issues based on tags to the appropriate repo(s).

Thoughts? Anybody want to work up relevant scripts for me?

**Comment by Maks3w** [comment]
I prefer the second option, label issues by component and attend those when a refactor is in the roadmap.

**Comment by froschdesign** [comment]
@Maks3w 
Matthew has given no options, these are 3 steps to solve the problem.

@weierophinney 
:+1:

**Comment by Maks3w** [comment]
On the split RFC I already mention this tool https://github.com/google/github-issue-mover for migrate issues.

Anyone can do it, no repo grants needed. Just if someone wants to do it or automatize it please add a comment for avoid duplicated efforts.

**Comment by GeeH** [comment]
I can take this on board @weierophinney once the docs migration is complete.

**Comment by adamlundrigan** [comment]
@Maks3w nice!  

@froschdesign I think @Maks3w was referring to the second option from my post:

> Manually review each open issue, closing the ones that are incomplete (not reproducible, insufficient detail, etc) and migrating the remaining ones to the correct repository.

I prefer this over Matthew's suggestion to bulk-close old issues with a "please reopen in the right place" comment as, realistically, very few are going to take the time to do that.  

How about this for a process flow?

  1. Volunteers go through all the open issues closing any that are obviously useless and tag the remaining ones with the relevant component tag(s). 

  2. Script auto-comments on each open issue pleading with the author to migrate the issue for us, but leave the issue open.  Perhaps we can deploy a custom version of github-issue-mover to make it REALLY simple for the OP to do the migration: pre-populate the issue link and destination repository fields (or provide a choice if issue is tagged with multiple components).  I can get this bit ready while I'm at TrueNorthPHP this week(end). 

  3. After suitable time has passed, volunteers and component maintainers start combing through the remaining open issues closing or migrating them on behalf of unresponsive OPs.  If volunteers are lacking then we can just bulk-move everything to the repository of the component it's tagged with. 

That should help move *some* of the burden of dealing with these open issues back onto the OPs who want to help.  Making it super easy for them to do so should hopefully help with uptake.

**Comment by froschdesign** [comment]
@adamlundrigan 
> I prefer this over Matthew's suggestion to bulk-close old issues with a "please reopen in the right place" comment as

Not "please reopen in the right place", but "open a new issue/PR on the relevant repo **if this is still a problem**"!

[The same way like in PHPUnit.](https://github.com/sebastianbergmann/phpunit/issues/1844)

**Comment by Maks3w** [comment]
I think is possible do that if the message include a text about this repo is maintained by volunteers and we appreciate the effort for check and open the issue again

**Comment by adamlundrigan** [comment]
@froschdesign yes, I was paraphrasing...a message similar to Sebastian's is definitely the right way, but we should also make it SUPER easy for the OP to migrate the issue themselves.  Just closing with a message like "please open a new ticket in the right place if this is still an issue" puts a lot of work on them to find the right repository then copy-and-paste their issue text, and IMO very few are going to do that.  Plus, if they don't mention the old ticket when creating the new one any conversation thread it contained will be lost and whomever picks up the new issue to triage it has to start again from scratch.

With a github-issue-mover setup like I mentioned the OP can move the issue with just two clicks, three if the issue is tagged with multiple components and they have to choose which repo it should go into.  One of those clicks is to authorize GitHub access, which they should only have to do once.

**Comment by froschdesign** [comment]
@adamlundrigan 
> we should also make it SUPER easy for the OP to migrate the issue themselves

If we can do that, then: :+1:

**Comment by adamlundrigan** [comment]
I've reviewed the open issues that didn't have a component label and added one where appropriate.  The following tickets are not tagged with a component as they concern the metapackage: 

 - [#7638](https://github.com/zendframework/zf2/issues/7638)

**Comment by GeeH** [comment]
After speaking with @adamlundrigan in person at True North and throwing the numbers around, here is what I propose:

1. We use a script to close any issues that haven't been updated this year with a message tagging the creator asking them to move the issue to the correct repository if the issue is still valid. We can link to the Google tool that Adam found to show them how to reopen the issue and migrate it in a few clicks. 

2. We'll be left with 251 open issues (minus this one), next step is to use a script to comment on those 251 issues asking the owner to migrate the issues using the tool outlined above if they think the issue is still valid, or to close it if they don't. Set a deadline on how long they have to do this (suggest 4 weeks).

3. Remaining issues - if there are a manageable number we can go through them by hand and close irrelevant ones and migrate any left open by hand using the Google tool. If there are still hundreds, I suggest we write a script that moves them to the tagged repo using the API and tags the original author. The downside is all the issues will be owned by a bot rather than the original owner but pragmatism > accuracy sometimes.

Thoughts?

**Comment by weierophinney** [comment]
@GeeH That's pretty much exactly what I had in mind. I think all of you have sufficient organization permissions to do this; let me know if you need any assistance from me, however!

**Comment by adamlundrigan** [comment]
Just a check-in to let you know that I haven't forgotten about this.  

I wanted to update github-issue-mover to make the process simpler -- feed in the issue ID via query string and it would pre-populate the fields, making the move process a one-click operation for the ticket owner.  That software is built with Dart and the google cloud engine SDK and I've had zero luck getting it to run locally so I can hack on it.  I've burned the time I had allocated to do that part.

Over the next day or two I'll draft up the issue comment that the script will post on each open ticket, including precise instructions on how to use the [public github-issue-mover instance](github-issue-mover.appspot.com) that @Maks3w found.  I'll post that message here once I have the first draft written.

**Comment by adamlundrigan** [comment]
Here's the first draft of the comment: https://gist.github.com/adamlundrigan/538324f8a2f8d9b76a84

**Comment by adamlundrigan** [comment]
I updated the Gist a few days ago with a draft comment to place on the PRs when the script auto-closes them.

**Comment by glen-84** [comment]
`take an active role` -> `taking an active role`
`to it's new home` -> `to its new home`

**Comment by adamlundrigan** [comment]
@glen-84 thanks, grammar fail on my part.  I've fixed that and reworded some lines that sound clunky now that several months have passed since I wrote it.

**Comment by adamlundrigan** [comment]
I rather spectacularly dropped the ball on this.  I'll be getting it back on track this month even if it kills me :shipit:

**Comment by GeeH** [comment]
Done.
